File: internal/export/export_test.go
package export

import (
	"strings"
	"testing"
	"time"

	"github.com/celados/vox/internal/dashscope"
	"github.com/celados/vox/internal/run"
	"gopkg.in/yaml.v3"
)

// word builds a word whose timing is caller-controlled, in milliseconds.
func word(text, punct string, begin, end int) dashscope.ASRWord {
	return dashscope.ASRWord{Text: text, Punctuation: punct, BeginTime: begin, EndTime: end}
}

func TestSegmentSplitsOnTerminator(t *testing.T) {
	cues := Segment([]dashscope.ASRWord{
		word("你好", "", 0, 500),
		word("世界", "。", 500, 1000),
		word("再见", "。", 1200, 1700),
	})
	if len(cues) != 2 {
		t.Fatalf("want 2 cues, got %d: %+v", len(cues), cues)
	}
	if cues[0].Text != "你好世界。" {
		t.Errorf("first cue = %q", cues[0].Text)
	}
	if cues[0].Begin.Milliseconds() != 0 || cues[0].End.Milliseconds() != 1000 {
		t.Errorf("first cue timing = %v..%v", cues[0].Begin, cues[0].End)
	}
	if cues[1].Index != 2 {
		t.Errorf("cue index = %d, want 2", cues[1].Index)
	}
}

// A silence longer than gapSplit ends a cue even with no punctuation, which is
// the only segmentation signal for speech the model punctuates sparsely.
func TestSegmentSplitsOnGap(t *testing.T) {
	cues := Segment([]dashscope.ASRWord{
		word("前半句", "", 0, 500),
		word("后半句", "", 1500, 2000),
	})
	if len(cues) != 2 {
		t.Fatalf("want 2 cues, got %d: %+v", len(cues), cues)
	}
}

func TestSegmentKeepsShortClausesTogether(t *testing.T) {
	// A separator below half the cue budget must not split.
	cues := Segment([]dashscope.ASRWord{
		word("好", "，", 0, 200),
		word("的", "。", 200, 400),
	})
	if len(cues) != 1 {
		t.Fatalf("want 1 cue, got %d: %+v", len(cues), cues)
	}
}

func TestSegmentEmpty(t *testing.T) {
	if cues := Segment(nil); len(cues) != 0 {
		t.Fatalf("want no cues, got %+v", cues)
	}
}

func TestTimeFormats(t *testing.T) {
	d := time.Hour + 2*time.Minute + 3*time.Second + 45*time.Millisecond
	if got := srtTime(d); got != "01:02:03,045" {
		t.Errorf("srtTime = %q", got)
	}
	if got := vttTime(d); got != "01:02:03.045" {
		t.Errorf("vttTime = %q", got)
	}
}

func TestRenderVTTHeader(t *testing.T) {
	out := renderVTT([]Cue{{Index: 1, End: time.Second, Text: "x"}})
	if !strings.HasPrefix(out, "WEBVTT\n\n") {
		t.Errorf("vtt output = %q", out)
	}
}

// Regression: a run with text but no word timings used to render an empty
// subtitle file and a markdown document containing only frontmatter.
func TestRendersWithoutWordTimings(t *testing.T) {
	rec := &run.Record{
		Meta:   run.Meta{SID: "abc", Size: run.Size{Duration: 12}},
		Result: &dashscope.ASRResult{Text: "整段文本，没有词级时间戳。"},
	}

	srt, err := Render(rec, "srt")
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(srt, "整段文本") {
		t.Errorf("srt lost the transcript: %q", srt)
	}
	if !strings.Contains(srt, "00:00:12,000") {
		t.Errorf("srt must span the run duration: %q", srt)
	}

	md, err := Render(rec, "md")
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(md, "整段文本") {
		t.Errorf("markdown lost the transcript: %q", md)
	}
}

// Regression: out-of-order or zero-length word timings produced cues whose end
// preceded their begin, which players reject.
func TestSegmentRepairsTimeline(t *testing.T) {
	cues := Segment([]dashscope.ASRWord{
		word("后", "。", 1000, 1500),
		word("先", "。", 500, 700), // arrives late but claims an earlier start
		word("零", "。", 2000, 2000),
	})
	prev := time.Duration(-1)
	for _, cue := range cues {
		if cue.End < cue.Begin {
			t.Errorf("cue %d: end %v precedes begin %v", cue.Index, cue.End, cue.Begin)
		}
		if cue.End == cue.Begin {
			t.Errorf("cue %d is zero-length", cue.Index)
		}
		if cue.Begin < prev {
			t.Errorf("cue %d starts before the previous cue ended", cue.Index)
		}
		prev = cue.End
	}
}

// Regression: the budget was checked after appending, so one long word could
// carry a cue far past the limits.
func TestSegmentKeepsCuesWithinBudget(t *testing.T) {
	var words []dashscope.ASRWord
	for i := 0; i < 40; i++ {
		words = append(words, word("字", "", i*100, i*100+100))
	}
	for _, cue := range Segment(words) {
		if runes := len([]rune(cue.Text)); runes > maxCueRunes {
			t.Errorf("cue %d holds %d runes, over the %d budget", cue.Index, runes, maxCueRunes)
		}
	}
}

// A YAML-significant character in the source path must not corrupt the
// frontmatter of an exported markdown transcript.
func TestMarkdownFrontmatterIsValidYAML(t *testing.T) {
	rec := &run.Record{
		Meta:   run.Meta{SID: "abc", Source: "/tmp/a: b #1.wav", Model: "m"},
		Result: &dashscope.ASRResult{Text: "x。"},
	}
	md := renderMarkdown(rec)

	_, body, found := strings.Cut(strings.TrimPrefix(md, "---\n"), "---\n")
	if !found {
		t.Fatalf("no frontmatter delimiters: %q", md)
	}
	head, _, _ := strings.Cut(strings.TrimPrefix(md, "---\n"), "---\n")

	var parsed map[string]any
	if err := yaml.Unmarshal([]byte(head), &parsed); err != nil {
		t.Fatalf("frontmatter is not valid YAML: %v\n%s", err, head)
	}
	if parsed["source"] != "/tmp/a: b #1.wav" {
		t.Errorf("source round-tripped as %v", parsed["source"])
	}
	if !strings.Contains(body, "x。") {
		t.Errorf("body = %q", body)
	}
}

File: src/export.test.ts
import { expect, test } from "bun:test";
import { parse } from "yaml";

import type { AsrWord } from "./dashscope.ts";
import {
  MAX_CUE_RUNES,
  render,
  renderMarkdown,
  renderVtt,
  segment,
  srtTime,
  vttTime,
} from "./export.ts";
import type { RunRecord } from "./run-store.ts";

function word(text: string, punct: string, begin: number, end: number): AsrWord {
  return { text, punctuation: punct, begin_time: begin, end_time: end };
}

test("segment splits on terminator", () => {
  const cues = segment([
    word("你好", "", 0, 500),
    word("世界", "。", 500, 1000),
    word("再见", "。", 1200, 1700),
  ]);
  expect(cues).toHaveLength(2);
  expect(cues[0]?.text).toBe("你好世界。");
  expect(cues[0]?.beginMs).toBe(0);
  expect(cues[0]?.endMs).toBe(1000);
  expect(cues[1]?.index).toBe(2);
});

test("segment splits on gap", () => {
  const cues = segment([word("前半句", "", 0, 500), word("后半句", "", 1500, 2000)]);
  expect(cues).toHaveLength(2);
});

test("segment keeps short clauses together", () => {
  const cues = segment([word("好", "，", 0, 200), word("的", "。", 200, 400)]);
  expect(cues).toHaveLength(1);
});

test("segment empty", () => {
  expect(segment([])).toHaveLength(0);
});

test("time formats", () => {
  const ms = (1 * 3600 + 2 * 60 + 3) * 1000 + 45;
  expect(srtTime(ms)).toBe("01:02:03,045");
  expect(vttTime(ms)).toBe("01:02:03.045");
});

test("render VTT header", () => {
  expect(
    renderVtt([{ index: 1, beginMs: 0, endMs: 1000, text: "x" }]).startsWith("WEBVTT\n\n"),
  ).toBe(true);
});

test("renders without word timings", () => {
  const rec: RunRecord = {
    digest: "d",
    meta: {
      sid: "abc",
      source: "x.wav",
      model: "m",
      created: "2026-08-01T00:00:00Z",
      path: "p",
      size: { tokens: 1, words: 1, chars: 1, duration: 12 },
    },
    args: { model: "m", format: "wav" },
    result: { text: "整段文本，没有词级时间戳。" },
  };
  const srt = render(rec, "srt");
  expect(srt).toContain("整段文本");
  expect(srt).toContain("00:00:12,000");
  expect(render(rec, "md")).toContain("整段文本");
});

test("segment repairs timeline", () => {
  const cues = segment([
    word("后", "。", 1000, 1500),
    word("先", "。", 500, 700),
    word("零", "。", 2000, 2000),
  ]);
  let prev = -1;
  for (const cue of cues) {
    expect(cue.endMs).toBeGreaterThanOrEqual(cue.beginMs);
    expect(cue.endMs).not.toBe(cue.beginMs);
    expect(cue.beginMs).toBeGreaterThanOrEqual(prev);
    prev = cue.endMs;
  }
});

test("segment keeps cues within budget", () => {
  const words: AsrWord[] = [];
  for (let i = 0; i < 40; i++) words.push(word("字", "", i * 100, i * 100 + 100));
  for (const cue of segment(words)) {
    expect([...cue.text].length).toBeLessThanOrEqual(MAX_CUE_RUNES);
  }
});

test("markdown frontmatter is valid YAML", () => {
  const rec: RunRecord = {
    digest: "d",
    meta: {
      sid: "abc",
      source: "/tmp/a: b #1.wav",
      model: "m",
      created: "2026-08-01T00:00:00Z",
      path: "p",
      size: { tokens: 1, words: 1, chars: 1, duration: 1 },
    },
    args: { model: "m", format: "wav" },
    result: { text: "x。" },
  };
  const md = renderMarkdown(rec);
  const stripped = md.replace(/^---\n/, "");
  const cut = stripped.indexOf("---\n");
  expect(cut).toBeGreaterThan(0);
  const head = stripped.slice(0, cut);
  const body = stripped.slice(cut + 4);
  const parsed = parse(head) as Record<string, unknown>;
  expect(parsed.source).toBe("/tmp/a: b #1.wav");
  expect(body).toContain("x。");
});
